File: stitch_agent/run/ci_parser.py
"""Parse CI configuration files into a list of executable CIJob objects.

Supports:
- GitLab CI (.gitlab-ci.yml)
- GitHub Actions (.github/workflows/*.y?ml)

For the MVP we parse top-level jobs, normalize script into a list of shell
commands, and respect stage order. Advanced features (includes, `needs:`,
matrix strategies, `uses:` action steps) are deferred.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

import yaml

from stitch_agent.run.ci_detect import CIPlatform
from stitch_agent.run.models import CIJob

if TYPE_CHECKING:
    from pathlib import Path


class CIParseError(Exception):
    """Raised when a CI configuration file cannot be parsed."""


# Reserved top-level keys in GitLab CI that are not jobs.
_GITLAB_RESERVED_KEYS: frozenset[str] = frozenset(
    {
        "default",
        "image",
        "include",
        "services",
        "stages",
        "variables",
        "workflow",
        "cache",
        "before_script",
        "after_script",
    }
)

_DEFAULT_GITLAB_STAGES: list[str] = [".pre", "build", "test", "deploy", ".post"]


def parse_ci_config(
    repo_root: Path,
    platform: CIPlatform = CIPlatform.UNKNOWN,
) -> list[CIJob]:
    """Parse CI files in the repo. Returns ordered jobs.

    When *platform* is a specific value (GITLAB or GITHUB), only that
    platform's config is parsed. When UNKNOWN, both are tried (original
    behaviour).

    Order: GitLab jobs first (if present), then GitHub workflow jobs sorted by
    filename. Within GitLab, jobs are ordered by stage. Within a stage, by
    insertion order from the YAML file.
    """
    jobs: list[CIJob] = []

    parse_gitlab = platform in (CIPlatform.GITLAB, CIPlatform.UNKNOWN)
    parse_github = platform in (CIPlatform.GITHUB, CIPlatform.UNKNOWN)

    if parse_gitlab:
        gl_path = repo_root / ".gitlab-ci.yml"
        if gl_path.is_file():
            jobs.extend(_parse_gitlab_ci(gl_path))

    if parse_github:
        gh_dir = repo_root / ".github" / "workflows"
        if gh_dir.is_dir():
            for wf in sorted(gh_dir.glob("*.yml")) + sorted(gh_dir.glob("*.yaml")):
                jobs.extend(_parse_github_workflow(wf))

    return jobs


def _load_yaml(path: Path) -> Any:
    try:
        return yaml.safe_load(path.read_text())
    except yaml.YAMLError as exc:
        raise CIParseError(f"Invalid YAML in {path}: {exc}") from exc


def _normalize_script(raw: Any) -> list[str]:
    """Convert a GitLab/GitHub script/run value into a list of shell commands."""
    if raw is None:
        return []
    if isinstance(raw, str):
        return [raw]
    if isinstance(raw, list):
        out: list[str] = []
        for item in raw:
            if isinstance(item, str):
                out.append(item)
            elif isinstance(item, list):
                # nested list: flatten
                out.extend(x for x in item if isinstance(x, str))
        return out
    return []


def _parse_gitlab_ci(path: Path) -> list[CIJob]:
    data = _load_yaml(path)
    if not isinstance(data, dict):
        return []

    stages_raw = data.get("stages")
    if isinstance(stages_raw, list) and stages_raw:
        stages_order = [s for s in stages_raw if isinstance(s, str)]
    else:
        stages_order = list(_DEFAULT_GITLAB_STAGES)

    default_image: str | None = None
    default_block = data.get("default")
    if isinstance(default_block, dict):
        img = default_block.get("image")
        if isinstance(img, str):
            default_image = img
        elif isinstance(img, dict) and isinstance(img.get("name"), str):
            default_image = img["name"]

    top_image = data.get("image")
    if default_image is None and isinstance(top_image, str):
        default_image = top_image
    elif (
        default_image is None
        and isinstance(top_image, dict)
        and isinstance(top_image.get("name"), str)
    ):
        default_image = top_image["name"]

    top_before_script = _normalize_script(data.get("before_script"))

    raw_jobs: list[tuple[str, dict[str, Any]]] = []
    for key, value in data.items():
        if not isinstance(key, str):
            continue
        if key.startswith("."):
            # hidden template
            continue
        if key in _GITLAB_RESERVED_KEYS:
            continue
        if not isinstance(value, dict):
            continue
        if "script" not in value and "run" not in value:
            continue
        raw_jobs.append((key, value))

    # Group by stage in insertion order, then output in stages_order.
    by_stage: dict[str, list[CIJob]] = {}
    for name, block in raw_jobs:
        stage = block.get("stage") if isinstance(block.get("stage"), str) else "test"
        if stage not in stages_order:
            stages_order = [*stages_order, stage]

        job_image: str | None = default_image
        img = block.get("image")
        if isinstance(img, str):
            job_image = img
        elif isinstance(img, dict) and isinstance(img.get("name"), str):
            job_image = img["name"]

        before = _normalize_script(block.get("before_script")) or top_before_script
        script = _normalize_script(block.get("script"))
        if not script:
            script = _normalize_script(block.get("run"))

        full_script = [*before, *script]
        job = CIJob(
            name=name,
            stage=stage or "test",
            script=full_script,
            image=job_image,
            source_file=path.name,
        )
        by_stage.setdefault(job.stage, []).append(job)

    ordered: list[CIJob] = []
    for stage in stages_order:
        ordered.extend(by_stage.get(stage, []))
    # Any stages not in stages_order (shouldn't happen since we extend above,
    # but defensive)
    for stage, stage_jobs in by_stage.items():
        if stage not in stages_order:
            ordered.extend(stage_jobs)
    return ordered


def _parse_github_workflow(path: Path) -> list[CIJob]:
    data = _load_yaml(path)
    if not isinstance(data, dict):
        return []

    jobs_block = data.get("jobs")
    if not isinstance(jobs_block, dict):
        return []

    stage = f"workflow:{path.name}"
    result: list[CIJob] = []
    for job_name, job_def in jobs_block.items():
        if not isinstance(job_name, str) or not isinstance(job_def, dict):
            continue

        steps = job_def.get("steps")
        script: list[str] = []
        if isinstance(steps, list):
            for step in steps:
                if not isinstance(step, dict):
                    continue
                run = step.get("run")
                if isinstance(run, str):
                    script.append(run)
                elif isinstance(run, list):
                    script.extend(x for x in run if isinstance(x, str))

        if not script:
            # Jobs with only `uses:` steps are skipped by the parser.
            continue

        image: str | None = None
        container = job_def.get("container")
        if isinstance(container, str):
            image = container
        elif isinstance(container, dict) and isinstance(container.get("image"), str):
            image = container["image"]

        result.append(
            CIJob(
                name=job_name,
                stage=stage,
                script=script,
                image=image,
                source_file=path.name,
            )
        )
    return result

File: src/core/ci-parser.ts
import { existsSync, readFileSync, readdirSync, statSync } from "node:fs";
import { join } from "node:path";
import yaml from "js-yaml";
import type { CIPlatform } from "./ci-detect.js";
import type { CIJob } from "./models.js";

export class CIParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CIParseError";
  }
}

const GITLAB_RESERVED_KEYS = new Set([
  "default",
  "image",
  "include",
  "services",
  "stages",
  "variables",
  "workflow",
  "cache",
  "before_script",
  "after_script",
]);

const DEFAULT_GITLAB_STAGES = [".pre", "build", "test", "deploy", ".post"];

export function parseCIConfig(repoRoot: string, platform: CIPlatform = "unknown"): CIJob[] {
  const jobs: CIJob[] = [];
  if (platform === "gitlab" || platform === "unknown") {
    collectGitlab(repoRoot, jobs);
  }
  if (platform === "github" || platform === "unknown") {
    collectGithub(repoRoot, jobs);
  }
  if (platform === "bitbucket" || platform === "unknown") {
    collectBitbucket(repoRoot, jobs);
  }
  return jobs;
}

function collectGitlab(repoRoot: string, jobs: CIJob[]): void {
  const glPath = join(repoRoot, ".gitlab-ci.yml");
  if (existsSync(glPath)) jobs.push(...parseGitlabCI(glPath));
}

function collectGithub(repoRoot: string, jobs: CIJob[]): void {
  const ghDir = join(repoRoot, ".github", "workflows");
  if (!existsSync(ghDir) || !statSync(ghDir).isDirectory()) return;
  const files = readdirSync(ghDir)
    .filter((f) => f.endsWith(".yml") || f.endsWith(".yaml"))
    .sort();
  for (const file of files) {
    jobs.push(...parseGithubWorkflow(join(ghDir, file)));
  }
}

function collectBitbucket(repoRoot: string, jobs: CIJob[]): void {
  const bbPath = join(repoRoot, "bitbucket-pipelines.yml");
  if (existsSync(bbPath)) jobs.push(...parseBitbucketPipelines(bbPath));
}

function loadYaml(path: string): unknown {
  try {
    const text = readFileSync(path, "utf-8");
    return yaml.load(text);
  } catch (err) {
    if (err instanceof yaml.YAMLException) {
      throw new CIParseError(`Invalid YAML in ${path}: ${err.message}`);
    }
    throw err;
  }
}

function normalizeScript(raw: unknown): string[] {
  if (raw === null || raw === undefined) return [];
  if (typeof raw === "string") return [raw];
  if (!Array.isArray(raw)) return [];
  return raw.flatMap((item) => {
    if (typeof item === "string") return item;
    if (Array.isArray(item)) return item.filter((x): x is string => typeof x === "string");
    return [];
  });
}

// ── GitLab CI ────────────────────────────────────────────────────────────

function resolveGitlabStages(doc: Record<string, unknown>): string[] {
  const raw = doc.stages;
  if (Array.isArray(raw) && raw.length > 0) {
    return raw.filter((s): s is string => typeof s === "string");
  }
  return [...DEFAULT_GITLAB_STAGES];
}

function resolveGitlabDefaultImage(doc: Record<string, unknown>): string | null {
  const defaultBlock = doc.default;
  if (typeof defaultBlock === "object" && defaultBlock !== null) {
    const img = extractImage((defaultBlock as Record<string, unknown>).image);
    if (img !== null) return img;
  }
  return extractImage(doc.image);
}

function extractGitlabRawJobs(doc: Record<string, unknown>): [string, Record<string, unknown>][] {
  const rawJobs: [string, Record<string, unknown>][] = [];
  for (const [key, value] of Object.entries(doc)) {
    if (key.startsWith(".") || GITLAB_RESERVED_KEYS.has(key)) continue;
    if (typeof value !== "object" || value === null) continue;
    const block = value as Record<string, unknown>;
    if ("script" in block || "run" in block) rawJobs.push([key, block]);
  }
  return rawJobs;
}

function buildGitlabJob(
  name: string,
  block: Record<string, unknown>,
  path: string,
  defaultImage: string | null,
  topBeforeScript: string[],
): CIJob {
  const jobImage = extractImage(block.image) ?? defaultImage;
  const beforeScript = normalizeScript(block.before_script);
  const before = beforeScript.length > 0 ? beforeScript : topBeforeScript;
  let script = normalizeScript(block.script);
  if (script.length === 0) script = normalizeScript(block.run);
  const stage = typeof block.stage === "string" ? block.stage : "test";
  return {
    name,
    stage,
    script: [...before, ...script],
    image: jobImage,
    sourceFile: path.split("/").pop() ?? "",
    skipReason: null,
  };
}

function orderByStage(byStage: Map<string, CIJob[]>, stagesOrder: string[]): CIJob[] {
  const ordered: CIJob[] = [];
  for (const stage of stagesOrder) {
    const jobs = byStage.get(stage);
    if (jobs) ordered.push(...jobs);
  }
  for (const [stage, jobs] of byStage) {
    if (!stagesOrder.includes(stage)) ordered.push(...jobs);
  }
  return ordered;
}

function parseGitlabCI(path: string): CIJob[] {
  const data = loadYaml(path);
  if (typeof data !== "object" || data === null) return [];
  const doc = data as Record<string, unknown>;

  const stagesOrder = resolveGitlabStages(doc);
  const defaultImage = resolveGitlabDefaultImage(doc);
  const topBeforeScript = normalizeScript(doc.before_script);
  const rawJobs = extractGitlabRawJobs(doc);

  const byStage = new Map<string, CIJob[]>();
  for (const [name, block] of rawJobs) {
    const job = buildGitlabJob(name, block, path, defaultImage, topBeforeScript);
    if (!stagesOrder.includes(job.stage)) stagesOrder.push(job.stage);
    if (!byStage.has(job.stage)) byStage.set(job.stage, []);
    byStage.get(job.stage)?.push(job);
  }

  return orderByStage(byStage, stagesOrder);
}

// ── GitHub Actions ───────────────────────────────────────────────────────

function extractGithubSteps(steps: unknown): string[] {
  if (!Array.isArray(steps)) return [];
  const script: string[] = [];
  for (const step of steps) {
    if (typeof step !== "object" || step === null) continue;
    const run = (step as Record<string, unknown>).run;
    if (typeof run === "string") {
      script.push(run);
    } else if (Array.isArray(run)) {
      script.push(...run.filter((x): x is string => typeof x === "string"));
    }
  }
  return script;
}

function extractGithubImage(def: Record<string, unknown>): string | null {
  const container = def.container;
  if (typeof container === "string") return container;
  if (typeof container === "object" && container !== null) {
    const img = (container as Record<string, unknown>).image;
    if (typeof img === "string") return img;
  }
  return null;
}

function parseGithubWorkflow(path: string): CIJob[] {
  const data = loadYaml(path);
  if (typeof data !== "object" || data === null) return [];
  const doc = data as Record<string, unknown>;

  const jobsBlock = doc.jobs;
  if (typeof jobsBlock !== "object" || jobsBlock === null) return [];

  const filename = path.split("/").pop() ?? "";
  const stage = `workflow:${filename}`;
  const result: CIJob[] = [];

  for (const [jobName, jobDef] of Object.entries(jobsBlock as Record<string, unknown>)) {
    if (typeof jobDef !== "object" || jobDef === null) continue;
    const def = jobDef as Record<string, unknown>;
    const script = extractGithubSteps(def.steps);
    if (script.length === 0) continue;
    result.push({
      name: jobName,
      stage,
      script,
      image: extractGithubImage(def),
      sourceFile: filename,
      skipReason: null,
    });
  }

  return result;
}

// ── Bitbucket Pipelines ──────────────────────────────────────────────────

function asRecord(val: unknown): Record<string, unknown> | null {
  if (typeof val === "object" && val !== null) return val as Record<string, unknown>;
  return null;
}

function extractBitbucketStep(entry: Record<string, unknown>): Record<string, unknown> | null {
  if ("step" in entry) return asRecord(entry.step);
  return null;
}

function walkParallelSteps(entry: Record<string, unknown>): Record<string, unknown>[] {
  const par = entry.parallel;
  const steps = Array.isArray(par) ? par : asRecord(par)?.steps;
  if (!Array.isArray(steps)) return [];
  const result: Record<string, unknown>[] = [];
  for (const ps of steps) {
    const rec = asRecord(ps);
    if (rec) {
      const inner = extractBitbucketStep(rec);
      if (inner) result.push(inner);
    }
  }
  return result;
}

function walkStageBlock(
  entry: Record<string, unknown>,
): { label: string; steps: Record<string, unknown>[] } | null {
  const stageBlock = asRecord(entry.stage);
  if (!stageBlock) return null;
  const label = typeof stageBlock.name === "string" ? stageBlock.name : "stage";
  const inner = stageBlock.steps;
  if (!Array.isArray(inner)) return { label, steps: [] };
  const steps: Record<string, unknown>[] = [];
  for (const s of inner) {
    const rec = asRecord(s);
    if (rec) {
      const step = extractBitbucketStep(rec);
      if (step) steps.push(step);
    }
  }
  return { label, steps };
}

interface BitbucketContext {
  defaultImage: string | null;
  filename: string;
  usedNames: Set<string>;
  result: CIJob[];
}

function pushBitbucketStep(
  ctx: BitbucketContext,
  stageLabel: string,
  step: Record<string, unknown>,
  fallbackIdx: number,
): void {
  const script = normalizeScript(step.script);
  if (script.length === 0) return;

  const name = typeof step.name === "string" ? step.name : `step-${fallbackIdx}`;
  let unique = name;
  let suffix = 2;
  while (ctx.usedNames.has(unique)) {
    unique = `${name}-${suffix++}`;
  }
  ctx.usedNames.add(unique);

  ctx.result.push({
    name: unique,
    stage: stageLabel,
    script,
    image: extractImage(step.image) ?? ctx.defaultImage,
    sourceFile: ctx.filename,
    skipReason: null,
  });
}

function collectEntrySteps(
  rec: Record<string, unknown>,
  stageLabel: string,
): { label: string; steps: Record<string, unknown>[] } {
  const direct = extractBitbucketStep(rec);
  if (direct) return { label: stageLabel, steps: [direct] };
  if ("parallel" in rec) return { label: stageLabel, steps: walkParallelSteps(rec) };
  if ("stage" in rec) {
    const block = walkStageBlock(rec);
    if (block) return { label: `${stageLabel}:${block.label}`, steps: block.steps };
  }
  return { label: stageLabel, steps: [] };
}

function walkPipelineList(ctx: BitbucketContext, stageLabel: string, list: unknown): void {
  if (!Array.isArray(list)) return;
  let idx = 0;
  for (const entry of list) {
    const rec = asRecord(entry);
    if (!rec) continue;
    const { label, steps } = collectEntrySteps(rec, stageLabel);
    for (const step of steps) {
      pushBitbucketStep(ctx, label, step, idx++);
    }
  }
}

function parseBitbucketPipelines(path: string): CIJob[] {
  const data = loadYaml(path);
  if (typeof data !== "object" || data === null) return [];
  const doc = data as Record<string, unknown>;

  const pipelines = doc.pipelines;
  if (typeof pipelines !== "object" || pipelines === null) return [];
  const pipelinesBlock = pipelines as Record<string, unknown>;

  const ctx: BitbucketContext = {
    defaultImage: extractImage(doc.image),
    filename: path.split("/").pop() ?? "",
    usedNames: new Set<string>(),
    result: [],
  };

  const simpleKinds = ["default"] as const;
  const keyedKinds = ["branches", "pull-requests", "tags", "custom"] as const;

  for (const kind of simpleKinds) {
    if (kind in pipelinesBlock) walkPipelineList(ctx, kind, pipelinesBlock[kind]);
  }
  for (const kind of keyedKinds) {
    const block = pipelinesBlock[kind];
    if (typeof block !== "object" || block === null) continue;
    for (const [pattern, list] of Object.entries(block as Record<string, unknown>)) {
      walkPipelineList(ctx, `${kind}:${pattern}`, list);
    }
  }

  return ctx.result;
}

// ── Shared ───────────────────────────────────────────────────────────────

function extractImage(raw: unknown): string | null {
  if (typeof raw === "string") return raw;
  if (typeof raw === "object" && raw !== null) {
    const r = raw as Record<string, unknown>;
    if (typeof r.name === "string") return r.name;
  }
  return null;
}
